### PR TITLE
Add Message and Event receivers

### PR DESCRIPTION
**Pull request by Zabuzard**
### Overview

This implements and closes long awaited #236 and #237 by adding two new interfaces `MessageReceiver` (+ `MessageReceiverAdapter`) and `EventReceiver` to the system.

Additionally, some refactoring had to be done to keep a good flow. First of all, `SlashCommand`, `MessageReceiver` and `EventReceiver` (and soon also `Routine`, see #235 ) are summarized under a marker interface called `Feature`. This marks all features that our bot core system supports.

Consequently, the previous starting class `Commands` has been renamed to `Features` and is now the origin of all features, not only commands (we already moved other stuff there in the past to prepare for this).

While at it, the existing class `RejoinMuteListener` has been adapted and now implements `EventReceiver` instead.

### Details

System-wise, `Features#createFeatures` returns a list of all features to `BotCore` and it will iterate and filter it based on the type (a sealed interface would be awesome here to not miss any future type).

For `SlashCommands`, nothing changed. `EventReceivers` are directly forwarded to `jda.addEventListener`, since that interface extends JDAs `EventListener`. `MessageReceivers` subscribe to channels they are interested in by a REGEX pattern matching channel names, `BotCore` takes care of the message event forwarding by iterating a simple `Map<Pattern, MessageReceiver>`.

### Examples

Quick example for a `MessageReceiver`, via its `MessageReceiverAdapter` helper:
```java
class Foo extends MessageReceiverAdapter {
  public Foo() {
    super(Pattern.compile("lobby"));
  }
  
  @Override
  void onMessageReceived(GuildMessageReceivedEvent event) {
    event.getMessage()
      .addReaction("🐸")
      .queue();
  }
}
```
a quick example for a `EventListener`:
```java
class Foo implements EventReceiver {
  @Override
  public void onEvent(GenericEvent event) {
    if (event instanceof GuildMessageReceivedEvent messageEvent) {
      System.out.println("Message was sent!");
    }
  }
}
```
and in both cases a simple addition in `Features.java`:
```java
features.add(new Foo());
```

### Remarks

When merging this, we have to go through our Wiki and do some minor adjusments caused by the renaming of `Commands` (now `Features`).

This PR competes against #341, which is a different design-approach to this problem, in which `SlashCommands` get the event listening capabilities directly instead of creating separate flows for them. (Which I am not a fan of, since this leads to creation of a behemoth class that can do everything instead of separating capabilities and concerns - SRP.)

### CODE REVIEW

**Comment by borgrel** [comment]
While this adds great functionality, it defeats the original design of the bot?

The bot has a SINGLE command listener and manually validates / splits all incoming slash commands to drastically reduce overhead (by only having the correct command respond to the event instead of letting each command manually filter out invalid commands they dont need to respond to)

The message system should work the same way? currently every message listener responds to all message events and have to manually filter the messages they are not interested in, instead of only receiving messages they do want.

File: application/src/main/java/org/togetherjava/tjbot/commands/Features.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/Features.java:57]: The routine support is added shortly after this PR, #346 , solving the valid issue you are mentioning.

So for the context of this PR, I would say we keep this for now.

File: application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java:97]: Agreed. So you are suggestion to remove this `onReady` method and instead let current users (probably only the help system) use the alternatives that are now available?

**Comment by Zabuzard** [comment]
> While this adds great functionality, it defeats the original design of the bot?

I am not sure I get your concern. I am in no way saying that people should use the new interfaces when all they want to do is to create a slash command. In particular, you should rarely have to use `EventReceiver` and instead use the alternatives `SlashCommand` and `MessageReceiver` if possible.

But it is also clear that there are a number of commands that do need more capabilities than just being a slash command (your help system is a good example for this). Hence why we need to introduce those features as well.
People can of course also combine things, `Foo implements SlashCommand, MessageReceiver`, if needed.

Or is your concern more about the fact that we have `jda.addEventListener(...)` on all the message/event receivers instead of letting them all go into `BotCore` and routing them ourselves? In that case, I was indeed also concidering this but at the end I couldn't see any benefit to this (other than introducing possible bugs). After all, why should we reinvent the existing routing logic inside JDA with our own if we have no benefit. Maybe you see benefit that I overlooked. For the slash commands there was clearly benefit since we had a need to decorate our slash command events with extra information.
The good thing is though that the detail of who routes the message/event receiver events is hidden inside `BotCore`, so it could be changed at any time if needed without impacting any user.

File: application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java
Comment (borgrel) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java:97]: no, the onready still has use / purpose

for example routine support could use it to log:
'INFO routine system loaded: 2 routines rolled back from invalid termination, 23 routines restarted, 5 old routines deleted'

we just need to remove the statement in the javadoc saying: 'use this to start things the bot doesnt support yet'

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java:97]: Routines are not slash commands though. Routines will use `Routine` (#346 ). So it should then rather be moved into that interface then.

But I get your point and agree, the Javadoc has to be adjusted.

**Comment by borgrel** [comment]
> Or is your concern more about the fact that we have `jda.addEventListener(...)` on all the message/event receivers instead of letting them all go into `BotCore` and routing them ourselves?

This

> The good thing is though that the detail of who routes the message/event receiver events is hidden inside `BotCore`, so it could be changed at any time if needed without impacting any user.

True enough, but not if we want to add internal routing


The fact is that few of our message listeners will want / need global message listening. for example the free system will only need to listen on the help channels. Illum's suggestion of a dad bot would only need to listen on lobby and geek speak. if the bot gets functionality that blocks ppl from talking in snippets (by deleting any posts after the first by the same user in a 24hr period) would only need to listen to messages in the snippets channel.

The way it currently is every single message in every single channel would be sent to every single listener and then every single listener has to filter out the 99% they dont need individually so assuming 12 listeners each message will be filtered 12 times and ignored 11 times, instead of being filtered once and only sent to the one listener that is interested in it.

The routing system is already designed and filters slash commands once to send them to the relevant command handler instead of just letting every command manually register its own listener and personally filter out the unneeded commands .... the routing should be expanded re-used to reduce overhead in message listeners in the same way

**Comment by Zabuzard** [comment]
> The fact is that few of our message listeners will want / need global message listening. for example the free system will only need to listen on the help channels. Illum's suggestion of a dad bot would only need to listen on lobby and geek speak. if the bot gets functionality that blocks ppl from talking in snippets (by deleting any posts after the first by the same user in a 24hr period) it would only need to listen to messages in the snippets channel.

Okay, I think I get your point now. You are suggessting to not only add a general `MessageReceiver` but a more specific way instead where people will also tell you which channels they are interested in, to only receive some messages instead.
So basically you are saying that majority (if not all) message receivers are only interested in a small portion of channels (or maybe even other filters, such as by author) and we should take care of that filtering to make it easier for users and also get rid of the code duplication that occurs when everyone has to filter themselves.

And if someone really has an interest in all channels, they could always also still fall back to `EventReceiver` which is seen as the unfiltered cumbersome fallback option.

I like that idea and will consider it, thanks.

File: application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java]: I removed `onReady` now. The free-system now implements `EventReceiver` (instead of `EventListener`), which also gets a `ReadyEvent` handed over. And now that the `BotCore` adds it to `jda.addEventListener(...)`, it also gets triggered early enough. Seems more elegant that way.

File: application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java
Comment (borgrel) [inline application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java:26]: Shouldnt this be `onMessageReceived`?

`onMessageSent` is confusing

**Review by borgrel — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java
Comment (borgrel) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java]: it looks like that will work.

technically what should happen is the JDA should be passed in on construction and the onReady code called in the constructor and then there is no need for generic events or the onready event (it was just being used as a hook to access the JDA without having to rewrite all of the code in slashadapter because commands didnt have access to JDA, Feature does.)

because the slashadapter and the messageadapter covers all functionality.

Comment (borgrel) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java]: but then again ...... maybe it should be left in `onReady`?
the Features construction is not threaded so heavy code there will block the bot's startup
and whether onReady is called by the JDA or by our bots router, it is threaded there.

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:300]: I'd personally we say, we implement EventListener and do instanceof checks ourselves.

Currently it goes from onEvent -> onGuildMessageUpdate -> onMessageUpdated, I don't think we need the ListenerAdapter

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:300]: Sounds reasonable.

File: application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/SlashCommand.java]: i feel like once this is merged, the free command can be simplified and made nicer but i leave that up to u for another PR. it would be too much for this PR to adjust /free beyond "just making it work", i think.

File: application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java]: done

**Comment by Zabuzard** [comment]
`MessageReceiver` now subscribe to channels they are interested in by their name and `BotCore` takes care of the forwarding (as requested by @borgrel ).

In a future PR, `/free` could now be simplified to basically implementing `MessageReceiver` and using that channel-name-subscription system, probably simplifying its own logic quite a bit (and getting rid of the `EventReceiver` part).

**Review by Tais993 — DISMISSED** [review]
It's good enough for now, it's blocking some other PR's (including #350)

We can change the subscription system for the Message events later one if required.

File: application/src/main/java/org/togetherjava/tjbot/commands/free/FreeCommand.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/free/FreeCommand.java:294]: I think we should put the whole `GuildMessageRevievedEvent` handling in a separate method. Why don't we just implement `MessageReciever` and have that block in the `onGuildMessageRecieved` method

File: application/src/main/java/org/togetherjava/tjbot/commands/free/UserStrings.java
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/free/UserStrings.java:26]: Maybe change this to `, please consult a moderator`?

File: application/src/main/java/org/togetherjava/tjbot/commands/free/ChannelMonitor.java
Comment (illuminator3) [inline application/src/main/java/org/togetherjava/tjbot/commands/free/ChannelMonitor.java:277]: New lines please.

**Review by illuminator3 — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/free/FreeCommand.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/free/FreeCommand.java:294]: I agree, but this should not be done in this PR but in a dedicated PR dealing with the free-command. borgrel is already awaitng this to get merged to do exactly those changes.

For this PR, i only did like the bare minimum to integrate the existing code so that it still works as before. Any further change would likely be a bigger refactoring and also needs more testing.

Comment (java-coding-prodigy) [reply application/src/main/java/org/togetherjava/tjbot/commands/free/FreeCommand.java:294]: I realized that from our discussion in the discord.

File: application/src/main/java/org/togetherjava/tjbot/commands/free/ChannelMonitor.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/free/ChannelMonitor.java:279]: Done, please check.

File: application/src/main/java/org/togetherjava/tjbot/commands/free/UserStrings.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/free/UserStrings.java]: Done, please check.

File: application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiver.java:32]:
```suggestion
    Predicate<AbstractChannel> getChannelPredicate();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiverAdapter.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/MessageReceiverAdapter.java:34]:
```suggestion
    public final @NotNull Predicate<AbstractChannel> getChannelPredicate() {
       return channel -> channelNamePattern.matcher(channel.getName()).matches();
  }
```

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:57]:
```suggestion
    private final Map<MessageReciever, Predicate<AbstractChannel>> channelNameToMessageReceiver = new HashMap<>();
```

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:76]:
```suggestion
                .put(messageReceiver, messageReciever.getChannelPredicate()));
```

File: application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java
Comment (java-coding-prodigy) [inline application/src/main/java/org/togetherjava/tjbot/commands/system/BotCore.java:154]:
```suggestion
            .filter(patternAndReceiver -> patternAndReceiver.getValue().test(channel))
            .map(Map.Entry::getKey);
```

**Review by java-coding-prodigy — DISMISSED** [review]
I have put my suggestions for the refactoring of our `MessageReciever` interface. However, you still have to change the imports, variable names, and documentation.

**Review by Tais993 — APPROVED** [review]

**Comment by Zabuzard** [comment]
7 days without real code changes + approval -> merging

**Comment by Zabuzard** [comment]
1 week without real code changes + 1 approval -> merging

**Pull request merged**
